### PR TITLE
Added flush interval to proxy to enable watch use through proxy

### LINKED ISSUE #3026: `osc get namespaces --watch` doesn't work through proxy
When running OpenShift against an external kubernetes, using watch commands against the OpenShift proxy doesn't work.

The web UI (using websockets) is working correctly, so it may be related to chunked transfer or caching.

### PR DESCRIPTION
**Pull request by stevekuznetsov**
Fixes #3026 

@liggitt @derekwaynecarr 

@brenton FYI

### CODE REVIEW

File: test/integration/external_kube_test.go [hunk 1]
```diff
@@ -20,6 +27,14 @@ func TestExternalKube(t *testing.T) {
 	if err != nil {
 		t.Fatalf("unexpected error: %v", err)
 	}
+	cluster1AdminKubeClient, err := testutil.GetClusterAdminKubeClient(cluster1AdminConfigFile)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	_, err = testutil.GetClusterAdminClient(cluster1AdminConfigFile)
```
Comment (liggitt) [inline test/integration/external_kube_test.go]: if this is unused, drop it

File: test/integration/external_kube_test.go [hunk 2]
```diff
@@ -55,13 +70,28 @@ func TestExternalKube(t *testing.T) {
...
-	_, err = testutil.StartConfiguredMasterWithOptions(cluster2MasterConfig, cluster2Options)
+	cluster2AdminConfigFile, err := testutil.StartConfiguredMasterWithOptions(cluster2MasterConfig, cluster2Options)
 	if err != nil {
 		t.Fatalf("unexpected error: %v", err)
 	}
+	cluster2AdminKubeClient, err := testutil.GetClusterAdminKubeClient(cluster2AdminConfigFile)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	_, err = testutil.GetClusterAdminClient(cluster2AdminConfigFile)
```
Comment (liggitt) [inline test/integration/external_kube_test.go]: if this is unused, drop it

File: test/integration/external_kube_test.go [hunk 3]
```diff
@@ -76,6 +106,46 @@ func TestExternalKube(t *testing.T) {
 	}
 }
 
+func watchProxyTest(cluster1AdminKubeClient, cluster2AdminKubeClient *kclient.Client, t *testing.T) {
+	// list namespaces in order to determine correct resourceVersion
+	namespaces, err := cluster1AdminKubeClient.Namespaces().List(labels.Everything(), fields.Everything())
+
+	// open a watch on Cluster 2 for namespaces starting with earliest resourceVersion
```
Comment (liggitt) [inline test/integration/external_kube_test.go]: s/earliest/latest/

File: test/integration/external_kube_test.go [hunk 4]
```diff
@@ -76,6 +106,46 @@ func TestExternalKube(t *testing.T) {
...
+	case e := <-namespaceWatch.ResultChan():
+		// check that the watch shows the new namespace
+		if e.Type != watch.Added {
+			t.Fatalf("expected an Added event but got: %v", e)
+		}
+		addedNamespace, ok := e.Object.(*kapi.Namespace)
+		if !ok {
+			t.Fatalf("unexpected cast error from event Object to Namespace")
+		}
+		if addedNamespace.ObjectMeta.Name != "test-namespace" {
```
Comment (liggitt) [inline test/integration/external_kube_test.go]: use `createdNamespace.Name`

File: pkg/util/httpproxy/upgradeawareproxy.go [hunk 5]
```diff
@@ -35,11 +36,13 @@ func NewUpgradeAwareSingleHostReverseProxy(clientConfig *kclient.Config, backend
 	if err != nil {
 		return nil, err
 	}
+	reverseProxy := httputil.NewSingleHostReverseProxy(backendAddr)
+	reverseProxy.FlushInterval = 50 * time.Millisecond
```
Comment (liggitt) [inline pkg/util/httpproxy/upgradeawareproxy.go]: looks like a super simple fix, which is nice. @smarterclayton @stefwalter is this a reasonable interval? I don't really know what the implications are

Comment (smarterclayton) [inline pkg/util/httpproxy/upgradeawareproxy.go]: I've seen similar

**Comment by stevekuznetsov** [comment]
@liggitt  fixed
[test]

File: pkg/util/httpproxy/upgradeawareproxy.go (hunk 5 above)
Comment (stefwalter) [inline pkg/util/httpproxy/upgradeawareproxy.go]: The interval is reasonable. The downside is that net/http/httputil seems to wake up every 50 milliseconds:

http://golang.org/src/net/http/httputil/reverseproxy.go?s=2744:2819#L207

Comment (liggitt) [inline pkg/util/httpproxy/upgradeawareproxy.go]: given that there are likely to be a lot of long-lived requests going through this proxy, I'm wondering if we should raise this to 100-200 ms

**Comment by stevekuznetsov** [comment]
[test]

**Comment by stevekuznetsov** [comment]
@liggitt Made the change to 200ms. Seems snappy still on my end.

**Comment by liggitt** [comment]
[merge]

**Pull request merged**